docs/dev/shimplify.md

```diff
@@ -0,0 +1,236 @@
+is called the *owner shim*.
+
+On the default read-only invocation path of the Maven build shimplify does not make any changes to
+shim source code files and their locations.
+
+* It analyzes the pre-shimplify directory structure and identifies the shims that through the code
+evolution ended up using more dedicated directories than necessary contributing avoidable
+complexity on top of an inherently complex directory structure already. As an example this is one of
+such warnings:
+
+```text
+shimplify - WARNING - Consider consolidating 312db, it spans multiple dedicated directories ['/home/user/gits/NVIDIA/spark-rapids/sql-plugin/src/main/312db/scala', '/home/user/gits/NVIDIA/spark-rapids/sql-plugin/src/main/31xdb/scala']
+```
+
+* For the shimplify directory structure all files under `src/(main|test)/spark*` directories
+are read to parse the `spark-rapids-shim-json-lines` comments. It performs the following
+validations:
+
+  * It makes sure that the comment is present and can be parsed
+  * The list of shims is non-empty (e.g., has not orphaned through dropping shims) and sorted.
+  * The file is stored under the *owner shim* directory.
+
+* All files participating listing the `buildver` of the current Maven build session are symlinked to
+`target/${buildver}/generated/src/(main|test)/(scala|java)`. Thus, instead of hardcoding distinct
+lists of directories for `build-helper` Maven plugin to add (one for each shim) after the full
+transition to shimplify, the pom will have only 4 add source statements that is independent of the
+number of supported shims.
+
+With the shimplify format in place it is easy to review all the files for a single shim without
+relying on Maven:
+
+```bash
+git grep '{"spark": "323"}' '*.scala' '*.java'
+```
+
+## Conversion to the Shimplify-based Directory Structure
+
+Shimplify can automatically convert the prior version-range-with-exceptions directory structure to
+the simplified version. This allows to make it an atomic transition without having to resolve
+almost unavoidable merge conflicts due to the sheer size of this sweeping change while the shim
+development is ongoing. The conversion of the shims source code and the regular build should not
+be done simultaneously for faster isolation and correction of potential bugs in the conversion code.
+
+Prior to invoking the conversion standalone, you first run
+
+```bash
+mvn clean install -DskipTests
+```
+
+on the current state of the `spark-rapids` repo.
+
+After that you can execute conversion in one or more iterations depending on specified -D parameters
+
+```bash
+mvn generate-sources antrun:run@shimplify-shim-sources -Dshimplify=true [-D...]
+```
+
+With `-Dshimplify=true`, shimplify is put on the write call path to generate and inject
+spark-rapids-shim-json-lines comments to all shim source files. The files are not yet moved to their
+owner shim directory, and so it is easy to verify with `git diff` the comments being injected. If
```

Comment:
As this tool matures, we may want to do this in a single step, so the default value for `-Dshimplify.move` will need to be true, may be not now but later.